Add allowed_mentions and mention everyone check to stream alerts

kablekompany opened the pull request:
Allows role mentions for discord.py 1.4+

### Type

- [x] Bugfix
- [ ] Enhancement
- [ ] New feature

### Description of the changes
Referencing line `747` in commit. Gives permissions to mention role on stream announcement.

kablekompany committed:
[streams]allowed_mentions on stream announcement

Allows role mentions for discord.py 1.4+
# Type

[x] Bugfix
[ ] Enhancement
[ ] New feature

Description of the changes

Referencing line 747. Gives permissions to mention role on stream announcement.

Jackenmen reviewed (commented):
While this PR solves the issue with mentions not actually pinging, do you think you could also make this not edit the roles if bot has mention everyone perms (as that perm now also allows to ping non-mentionable roles)?

kablekompany committed:
Update streams.py

kablekompany committed:
Add `mention_everyone` defaults

Lines for reference:
- [751-753]:
```py
if can_mention_everyone:
  await self.save_streams()
  return # if bot can mention everyone already, let's stop here```
- [722-774]: 
```py
if can_mention_everyone:
  mentions.append(role.mention)
  return # if bot can mention everyone already, let's stop here (part two)```

Hopefully, this is what you had in mind? Humbly admit I may not have a business in biting this piece but looking forward to feedback on the best methodology for the wanted outcome.
Thanks!

kablekompany committed:
*sweats*

kablekompany committed:
imagine variable naming

Should complete all passes.

kablekompany committed:
misplaced unindent on return

kablekompany commented:
## Lines for reference:

- `[750-754]`:
```py
guild = channel.guild
can_mention_everyone = guild.me.guild_permissions.mention_everyone
if can_mention_everyone:
  await self.save_streams()
  return # if bot can mention everyone already, let's stop here 
```

- `[772-776]`: 
```py
can_mention_everyone = guild.me.guild_permissions.mention_everyone
if can_mention_everyone:
  for role in guild.roles:
    mentions.append(role.mention)
    return  # if bot can mention everyone already, let's stop here
```


Hopefully, this is what you had in mind? Humbly admit I may not have a business in biting this piece but looking forward to feedback on the best methodology for the wanted outcome. Sorry for the redraws on simple mistakes...

Thanks!

Jackenmen commented on redbot/cogs/streams/streams.py:
```diff
@@ -741,8 +741,17 @@ async def check_streams(self):
+                        can_mention_everyone = guild.me.guild_permissions.mention_everyone
+                        if can_mention_everyone:
+                            await self.save_streams()
+                            return
```
This can all be removed if you apply my other suggestion.
```suggestion
```

Jackenmen commented on redbot/cogs/streams/streams.py:
```diff
@@ -760,6 +769,11 @@ async def _get_mention_str(self, guild: discord.Guild) -> Tuple[str, List[discor
+        if can_mention_everyone:
+            for role in guild.roles:
+                mentions.append(role.mention)
+                return
```
This can be all removed if you apply my other suggestion.
```suggestion
```

Jackenmen commented on redbot/cogs/streams/streams.py:
```diff
@@ -760,6 +769,11 @@ async def _get_mention_str(self, guild: discord.Guild) -> Tuple[str, List[discor
+                return
         for role in guild.roles:
             if await self.config.role(role).mention():
                 if can_manage_roles and not role.mentionable:
```
This is all that needs to be done here:
```suggestion
                if not can_mention_everyone and can_manage_roles and not role.mentionable:
```

Jackenmen commented on redbot/cogs/streams/streams.py:
```diff
@@ -760,6 +769,11 @@ async def _get_mention_str(self, guild: discord.Guild) -> Tuple[str, List[discor
         if await settings.mention_here():
             mentions.append("@here")
         can_manage_roles = guild.me.guild_permissions.manage_roles
+        can_mention_everyone = guild.me.guild_permissions.mention_everyone
```
It would be better if this were checking for permissions in channel:
```suggestion
        can_mention_everyone = channel.permissions_for(guild.me).mention_everyone
```
This requires you to also add a `channel` argument to this function and pass it in the place where it's called.

Jackenmen reviewed (changes requested):
The added changes make the bot mention all roles in the guild, which is not the same behavior as there was before.

This can also be simplified a lot, I sent the review comments which should make it clear what should be done here.

kablekompany committed:
Update redbot/cogs/streams/streams.py

Co-authored-by: jack1142 <6032823+jack1142@users.noreply.github.com>

kablekompany committed:
Update redbot/cogs/streams/streams.py

Co-authored-by: jack1142 <6032823+jack1142@users.noreply.github.com>

kablekompany committed:
Update redbot/cogs/streams/streams.py

Co-authored-by: jack1142 <6032823+jack1142@users.noreply.github.com>

kablekompany committed:
You're a beast jack

Co-authored-by: jack1142 <6032823+jack1142@users.noreply.github.com>

kablekompany committed:
this would cover all channels in guild (think)

kablekompany committed:
give this a reeeeee

kablekompany commented:
committed suggestions and added channel iteration for `can_mention_everyone` https://github.com/kablekompany/Red-DiscordBot/blob/8eb21913bb402b634af7a22a0414ede7c3014eb4/redbot/cogs/streams/streams.py#L767

Jackenmen commented on redbot/cogs/streams/streams.py:
```diff
@@ -760,9 +764,11 @@ async def _get_mention_str(self, guild: discord.Guild) -> Tuple[str, List[discor
         if await settings.mention_here():
             mentions.append("@here")
         can_manage_roles = guild.me.guild_permissions.manage_roles
+        for channel in guild.channels:
+            can_mention_everyone = channel.permissions_for(guild.me).mention_everyone
```
As I said before, you need to add `channel` param to `_get_mention_str()` method so that the bot checks the permission for the channel it will send the message in. Iterating through all channels here doesn't make much sense - we only care about permissions in *that specific channel* the message will be sent in.

kablekompany replied:
understood. param added, iteration removed

kablekompany committed:
Update streams.py

Jackenmen reviewed (changes requested):
Looks alright now, there's just one thing left - it looks like you forgot to pass the channel argument to `_get_mention_str()` call in line 716.

kablekompany committed:
adds channel argument

https://github.com/Cog-Creators/Red-DiscordBot/pull/4182#pullrequestreview-468012281

kablekompany committed:
image style checking, self

kablekompany committed:
imagine

kablekompany commented:
big time appreciate your patience and guidance @jack1142

Jackenmen reviewed (approved):
Looks good to me, and thanks so much for including new functionality related to Mention Everyone perm as well ❤️
